doc/source/example-lm.rst

```diff
@@ -0,0 +1,298 @@
+
+Two main componets of ``ray_train.py`` are a ``RayDistributedActor`` class and a new main function ``ray_main()``. The ``RayDistributedActor`` sets proper arguments for different ray actor processes, adds a checkpoint hook to enable the process to make use of new available GPUs, and call the ``main`` of Fairseq:
+
+.. code-block:: python
+
+  import math
+  import copy
+  import socket
+  import time
+
+  import ray
+
+  import fairseq
+  from fairseq import options
+  from fairseq_cli.train import main
+  from contextlib import closing
+
+  _original_save_checkpoint = fairseq.checkpoint_utils.save_checkpoint
+
+
+  class RayDistributedActor:
+      """
+      Ray distributed actor to perform distributed training.
+      """
+      def run(self, url, world_rank, args):
+          """
+          Set different fields in args for different ray actor processes, add a
+          checkpoint hook, and call the main function of fairseq.
+          """
+
+          # Set the init_method and rank of the process for distributed training.
+          print("Ray worker at {url} rank {rank}".format(url=url, rank=world_rank))
+          self.url = url
+          self.world_rank = world_rank
+          args.distributed_rank = world_rank
+          args.distributed_init_method = url
+
+          # Add a hook to the original save_checkpoint function to check whether
+          # or not there is new computational resources available. If so, raise
+          # an exception to restart the training process and make use of the new
+          # resources.
+          if args.cpu:
+              original_n_cpus = args.distributed_world_size
+
+              def _new_save_checkpoint(*args, **kwargs):
+                  _original_save_checkpoint(*args, **kwargs)
+                  n_cpus = int(ray.cluster_resources()["CPU"])
+                  if n_cpus > original_n_cpus:
+                      raise Exception("New CPUs find (original %d CPUs, now %d CPUs)"
+                                      % (original_n_cpus, n_cpus))
+          else:
+              original_n_gpus = args.distributed_world_size
+
+              def _new_save_checkpoint(*args, **kwargs):
+                  _original_save_checkpoint(*args, **kwargs)
+                  n_gpus = int(ray.cluster_resources().get("GPU", 0))
+                  if n_gpus > original_n_gpus:
+                      raise Exception("New GPUs find (original %d GPUs, now %d GPUs)"
+                                      % (original_n_gpus, n_gpus))
+          fairseq.checkpoint_utils.save_checkpoint = _new_save_checkpoint
```

Comment:
Can you put this into a helper function that passes in args?

Then, we can add the comment as a separate code block.

Reply:
resolved in the latest comment